streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -692,9 +688,10 @@ private void closeRecordCollector(final boolean clean) {
     @Override
     public void addRecords(final TopicPartition partition, final Iterable<ConsumerRecord<byte[], byte[]>> records) {
         if (state() == State.CLOSED || state() == State.CLOSING) {
-            log.info("Stream task {} is already closed, probably because it got unexpectedly migrated to another thread already. " +
-                         "Notifying the thread to trigger a new rebalance immediately.", id());
-            throw new TaskMigratedException(id());
+            // a task is only closing / closed when 1) task manager is closing, 2) a rebalance is undergoing;
+            // in either case we can just log it and move on without notifying the thread since the consumer
+            // would soon be updated to not return any records for this task anymore.
+            log.info("Stream task {} is already in {} state, skip adding records to it.", id(), state());
```

Comment:
Hm. I think we should actually be concerned if we ever get to here -- I'm not sure the `TaskMigratedException` made sense either. Trying to add records to a closed task would imply that we closed the task due to shutting down or because we no longer own it, both cases which should also involves trimming its topic(s) from the consumer's assignment, but were still returned records for said topic(s).

Unless, the consumer may still return already-fetched records from partitions no longer in its assignment during poll? I thought we would trim those records out and only return from the actual assignment

Reply:
Here's my rationale:

If the task is closed due to rebalance (i.e. we `#handleAssignment` or `#handleLostAll`), there might still be some buffered records from the consumer that are returning (since we update the subscription of consumer after), in this case since the subscription would be updated in the next iteration and no records would be returned, it is okay to just skip this once.

If the task is closed due to closing the thread, then there's no need to throw an exception either.